### PR TITLE
feat: support to create table from record batch iterator

### LINKED ISSUE #1562: bug(python): `create_table()` can't take a generator as data parameter when connect to remote database
### LanceDB version

v0.12.0

### What happened?
This is from [discord report](https://discord.com/channels/1030247538198061086/1105226910528716912/1277230905584259185)

Get error when the try to create table with `data=some_data_generator` in remote database:
```
File ~/miniconda3/envs/hl-reposearch/lib/python3.11/site-packages/lancedb/remote/arrow.py:20, in to_ipc_binary(table)
     18 """Serialize a PyArrow Table to IPC binary."""
...
---> 20 with pa.ipc.new_stream(sink, table.schema) as writer:
     21     writer.write_table(table)
     22 return sink.getvalue().to_pybytes()

AttributeError: 'generator' object has no attribute 'schema'
```
This happens only if connected to a remote database.

### Are there known steps to reproduce?

```
table = db.create_table(
    lance_db_table_name,
    get_batches_generator(),
    schema=Item.to_arrow_schema(),
)

def get_batches_generator():
    result = []
    batches = [nodes[i : i + batch_size] for i in range(0, len(nodes), batch_size)]
    for i, batch in enumerate(batches):
        yield pd.DataFrame.from_records(batch)

class Item(LanceModel):  
    vector: Vector(func.ndims()) = func.VectorField()
    content: str = func.SourceField() 
    file_path: str
    file_name: str
```

### PR DESCRIPTION
**Pull request by BubbleCal**
resolve #1562
- fix that RemoteTable can't create table from iterator
- infer the schema from the iterator
- refactor the checks for creating table

### CODE REVIEW

File: python/python/lancedb/table.py [hunk 1]
```diff
@@ -117,15 +117,59 @@ def _sanitize_data(
...
+            schema = data.schema
+        elif isinstance(data, Iterable):
+            if metadata:
+                raise TypeError(
+                    (
+                        "Persistent embedding functions not yet "
+                        "supported for generator data input"
+                    )
+                )
+        data = pa.Table.from_pylist([], schema=schema)
```
Comment (wjones127) [inline python/python/lancedb/table.py:165]: small nit: it's faster to not go through the Python -> Arrow conversion methods if you just want to create an empty table.

```suggestion
        data = pa.Table.from_batches([], schema=schema)
```

```python
In [5]: %timeit pa.Table.from_pylist([], schema)
6.1 µs ± 519 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [6]: %timeit pa.Table.from_batches([], schema)
287 ns ± 4.87 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

File: python/python/lancedb/table.py [hunk 2]
```diff
@@ -117,15 +117,59 @@ def _sanitize_data(
...
+        elif hasattr(data, "schema"):
+            schema = data.schema
+        elif isinstance(data, Iterable):
+            if metadata:
+                raise TypeError(
+                    (
+                        "Persistent embedding functions not yet "
+                        "supported for generator data input"
+                    )
+                )
```
Comment (wjones127) [inline python/python/lancedb/table.py:164]: I'm confused. In this branch, `data is None`, right? So neither of these will ever apply?

**Review by wjones127 — COMMENTED** [review]
Some initial questions

File: python/python/lancedb/table.py (hunk 1 above)
Comment (BubbleCal) [reply python/python/lancedb/table.py]: Fixed

File: python/python/lancedb/table.py (hunk 2 above)
Comment (BubbleCal) [reply python/python/lancedb/table.py]: Fixed

**Review by wjones127 — APPROVED** [review]

**Pull request merged**